torch/csrc/jit/codegen/cuda/lower_predicate.cpp

```diff
@@ -175,6 +177,488 @@ std::vector<kir::Expr*> generateConditionalFromPredicate(
+                [](IterDomain* axis) {
+                  return axis->getParallelType() ==
+                      ParallelType::MisalignedVectorize;
+                })) {
+          return true;
+        }
+      }
+    }
+    return false;
+  });
+
+  // Shift is not supported yet.
+  filters.push_back([](Expr* expr) {
+    auto& halo_info = GpuLower::current()->haloInfo();
+    auto input_tvs = ir_utils::filterByType<TensorView>(expr->inputs());
+    return halo_info.needsShiftPredicate(expr) ||
+        std::any_of(input_tvs.begin(), input_tvs.end(), [&](auto input_tv) {
+             return input_tv->definition() != nullptr &&
+                 halo_info.needsShiftPredicate(input_tv->definition());
+           });
+  });
+
+  // Predicates the expression if any producer-consumer pair of the
+  // expression needs to be predicated
+  filters.push_back([](Expr* expr) {
+    for (auto output : ir_utils::filterByType<TensorView>(expr->outputs())) {
+      for (auto input : ir_utils::filterByType<TensorView>(expr->inputs())) {
+        if (PredicateAnalyzer::needsPredicate(input, output)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  });
+
+  // Predicates Welford ops
+  filters.push_back([](Expr* expr) { return expr->isA<WelfordOp>(); });
+
+  // If this is a reduction, and if we omit the predicate for the
+  // input, the input may have a garbabe value, which must not be used
+  // for this reduction. However, if the input is also an output of
+  // another reduction with the same binary op, which is a common
+  // pattern with rfactor, the input should be safe to use with no
+  // predication.
+  filters.push_back([this](Expr* expr) {
+    if (expr->isA<ReductionOp>()) {
+      auto input = expr->inputs()[0]->as<TensorView>();
+      auto input_def = input->definition();
+      // When input_def is null, input must be an input to the fusion,
+      // so that must be allocated on global memory. Since we don't omit
+      // predication for expressions involving global memory, this
+      // should never occur.
+      TORCH_INTERNAL_ASSERT(
+          input_def != nullptr, "Inconsistent input found: ", input);
+
+      if (non_predicated_exprs_.find(input_def) !=
+              non_predicated_exprs_.end() &&
+          !(input_def->isA<ReductionOp>() &&
+            (expr->as<ReductionOp>()->getReductionOpType() ==
+             input_def->as<ReductionOp>()->getReductionOpType()))) {
+        return true;
+      }
+    }
+    return false;
+  });
+
+  // If any of the filters returns true, predicate must be used.
+  return std::any_of(filters.begin(), filters.end(), [expr](auto filter) {
+    return filter(expr);
+  });
+}
+
+void PredicateElimination::handle(Expr* expr) {
+  if (!ir_utils::isTVOp(expr)) {
+    return;
+  }
+
+  if (needsPredicate(expr)) {
+    return;
+  }
+
+  non_predicated_exprs_.insert(expr);
+
+  // Ensure all inputs have some values set at the out-of-bound
+  // regions
```

Comment:
Why do we care for anything not going into a reduction?

Reply:
I'm a little unclear on why you'll set initial values to nullptr as it seems setting that doesn't do anything. It doesn't seem like you're initializing values unless it comes from "back propagating" a reduction which should have an actual value to set.

Reply:
`nullptr` is just used as a placeholder to indicate the out-of-bound of a tensor should be initialized with "some" value, but we don't care what value to use. 

This isn't technically necessary, but I thought that might be better to be initialized anyway so that we never use undefined values. Something like compute-sanitizer may complain, though I don't think compute-sanitizer checks register usage.